multisense_ros/src/ground_surface_utilities.cpp

```diff
@@ -0,0 +1,279 @@
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+
+#include <multisense_ros/ground_surface_utilities.h>
+
+namespace ground_surface_utilities {
+
+namespace {
+
+float computeAzimuth(const float a, const float b)
+{
+    return a >= 0 ? static_cast<float>(M_PI) - atan(b / (a + std::numeric_limits<float>::epsilon()))
+                  : atan(b / -(a + std::numeric_limits<float>::epsilon()));
+}
+
+float computeRange(const float a, const float b)
+{
+    return sqrt(pow(a, 2) + pow(b, 2));
+}
+
+template <typename T> T computeQuadraticSurface(T x, T z, const T* params)
+{
+    ///
+    /// Compute height of quadratic surface model
+    ///     y = ax^2 + bz^2 + cxz + dx + ez + f
+    ///
+    return params[0] * pow(x, 2) + params[1] * pow(z, 2) + params[2] * x * z +
+           params[3] * x + params[4] * z + params[5];
+}
+
+Eigen::Matrix<Eigen::Matrix<float, Eigen::Dynamic, 1>, Eigen::Dynamic, 1> generateBasisArray()
+{
+    Eigen::Matrix<Eigen::Matrix<float, Eigen::Dynamic, 1>, Eigen::Dynamic, 1> ret(4);
+
+    // Temporary storage for polynomial coefficients
+    Eigen::Matrix<float, Eigen::Dynamic, 1> basisCoefficients(4);
+
+    // The following basis functions are copied from Lee, Wolberg, and Shin,
+    // "Scattered Data Interpolation with Multilevel B-Splines",
+    // IEEE Transactions on Visualization and Computer Graphics, Vol 3, 228-244, 1997.
+
+    // First cubic spline basis component is
+    // B(s) = (1 - s)**3 / 6.
+    // This expands to
+    // B(s) = -(1/6)s**3 + (1/2)s**2 - (1/2)s + 1/6.
+    basisCoefficients(0) = 1.0 / 6.0;
+    basisCoefficients(1) = -0.5;
+    basisCoefficients(2) = 0.5;
+    basisCoefficients(3) = -1.0 / 6.0;
+    ret(0) = basisCoefficients;
+
+    // Second cubic spline basis component is
+    // B(s) = (1/2)t**3 - t**2 + 2/3.
+    basisCoefficients(0) = 2.0 / 3.0;
+    basisCoefficients(1) = 0.0;
+    basisCoefficients(2) = -1.0;
+    basisCoefficients(3) = 0.5;
+    ret(1) = basisCoefficients;
+
+    // Third cubic spline basis component is
+    // B(s) = -(1/2)t**3 + (1/2)t**2 + (1/2)t + 1/6.
+    basisCoefficients(0) = 1.0 / 6.0;
+    basisCoefficients(1) = 0.5;
+    basisCoefficients(2) = 0.5;
+    basisCoefficients(3) = -0.5;
+    ret(2) = basisCoefficients;
+
+    // Fourth cubic spline basis component is
+    // B(s) = (1/6)t**3.
+    basisCoefficients(0) = 0.0;
+    basisCoefficients(1) = 0.0;
+    basisCoefficients(2) = 0.0;
+    basisCoefficients(3) = 1.0 / 6.0;
+    ret(3) = basisCoefficients;
+
+    return ret;
+}
+
+template <class FloatT>
+FloatT getSplineValue(
```

Comment:
You should move this template definition to the header file

Reply:
Just checking in about this - `getSplineValue` is only used inside `ground_surface_utilities.cpp` and is in an anonymous namespace. Basically I was trying to abstract it away and only expose the top-level functions that are required to draw the spline (see `multisense/multisense_ros/include/multisense_ros/ground_surface_utilities.h`).

But, if you think it's still best to move to the header file I can totally do that.

Reply:
Gotcha. That should be fine then.